keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1363 @@
+server side apply merge algorithm is run.
+
+This solves the vast majority of value removal needs. Specifically:
+
+| Schema type | Merge type | Example of how to unset a value                                          |
+|-------------| -----------|--------------------------------------------------------------------------|
+| struct      | atomic     | ```Object{ spec: Object.spec{ structField: {?fieldToRemove: optional.none()}}}``` |
+| struct      | granular   | `?fieldToRemove: optional.none()`                                        |
+| map         | atomic     | `mapField: object.spec.mapField.filter(k, k != "keyToRemove")`           |
+| map         | granular   | `mapField: {?"keyToRemove": optional.none()}`                            |
+| list        | atomic     | `listField: object.spec.listField.filter(e, e != "itemToRemove")`        |
+| list        | set        | `setField: object.spec.setField.filter(e, e != "itemToRemove")`          |
+| list        | map        | See below                                                                |
+| list        | granular   | Not Supported                                                            | 
+
+
+List with "map" merge type:
+  - Filter `objects.filter(<list>, <keys-to-remove>)` could be used for the deletion
+  - For associatedList with multiple keys like example above, a directive field added could be used to indicate the deletion.
+```yaml
+mutations:
+  - mutaton: >
+      Object{
+        spec: Object.spec{
+                assocListField: [Object.spec.assocListField{
+                       keyField1: "key1",
+                       keyField2: "key2",
+                       _: optional.none()
+                }]
+        }
+}
+
+```
+
+For granular list removal, an use case would be removing an entire container with an envvar named `remove-this-container`.
+```yaml
+mutations:
+  - mutation: >
+      Object{
+        spec: Object.spec{
+                containers: [Object.spec.containers{
+                       envvar: "remove-this-container",
+                       _: optional.none()
+                }]
+        }
+      }
+```
+It will traverse the container list and remove item whose envvar matches above. If multiple fields specified, the item needs to match all fields in order to be cleared.
+Notes: 
+- It will not do exact match for list item hence it will remove all item matches the specified field.
+- Multiple items could be cleared if all matches the field specified. 
+<<UNRESOLVED>>Deletion for primivates list<<UNRESOLVED>>
+
+##### Safety
+
+To ensure mutations are not "broken" by other mutations (overwritten, undone, or
+otherwise invalidated) and ensure the deterministic final state due to the random running order, \
+we provide an option to check if rerun certain mutation policy leads to object change.
+
+These validation checks will be declared using a `mutationValidationPolicy` field, which is an enum of the following values:
```

Comment:
is this for during the second invocation of the mutating admission policy or after the second invocation of the entire mutating admission chain.  I think these only make sense for after the second invocation of the entire chain, since a mutating admission policy may not be able to predict whether a new item was added to a slice.  Take the example of setting imagePullPolicy to Always.  We definitely want to re-run (not skip), but we also don't want to warn.

But running this after the second invocation, I could see wanting to ensure no mutation happened.  But in that case I don't understand why Ignore is grouped with Warn.

Reply:
Thanks for the comment! Yes it should happen after invocation and it will help to ensure a idempotent mutation is written because of random ordering. The ignore was in there because it didn't affect the request interception even with warning returned. I have removed ignore to reduce confusion :)

Reply:
> Yes it should happen after invocation

After the *second* invocation?  Perhaps framed differently: if set to Fail or Warn, will this run as part of the validating admission chain?